## The Electronic Nose Dataset for Plant Oil with Different Compositions

In [10]:
import os
import pandas as pd
from pathlib import Path

# Define the dataset path
dataset_path = Path('/Users/momthu/Documents/duhoc/Hungary/elte/research/oil/dataset_Oils')

# Get all oil type folders
oil_folders = sorted([d for d in dataset_path.iterdir() if d.is_dir()])
print(f"Found {len(oil_folders)} oil type folders:")
for folder in oil_folders:
    print(f"  - {folder.name}")


Found 19 oil type folders:
  - CO
  - COO
  - COO1
  - COO2
  - COO3
  - COO4
  - COO5
  - RO0
  - RO1
  - RO2
  - RO3
  - RO4
  - RO5
  - SO0
  - SO1
  - SO2
  - SO3
  - SO4
  - SO5


In [11]:
# Check columns consistency across all files
columns_dict = {}
total_files = 0
files_with_issues = []

for oil_folder in oil_folders:
    txt_files = sorted(oil_folder.glob('*.txt'))
    print(f"\n{oil_folder.name}: {len(txt_files)} files")
    
    if txt_files:
        # Read first file to get columns
        first_file = txt_files[0]
        df = pd.read_csv(first_file)
        columns = tuple(df.columns)
        
        if columns not in columns_dict:
            columns_dict[columns] = []
        
        columns_dict[columns].append((oil_folder.name, first_file.name, len(txt_files)))
        print(f"  Columns: {list(df.columns)}")
        print(f"  Shape: {df.shape}")
        total_files += len(txt_files)
        
        # Check if all files in this folder have the same columns
        for txt_file in txt_files[1:]:
            df_check = pd.read_csv(txt_file)
            if tuple(df_check.columns) != columns:
                files_with_issues.append((oil_folder.name, txt_file.name))

print(f"\n{'='*60}")
print(f"TOTAL FILES: {total_files}")
print(f"{'='*60}")


CO: 70 files
  Columns: ['Time [s]', 'TGS2620', 'TGS2602', 'TGS2603', 'TGS2610', 'TGS2611', 'TGS2600']
  Shape: (150, 7)

COO: 70 files
  Columns: ['Time [s]', 'TGS2620', 'TGS2602', 'TGS2603', 'TGS2610', 'TGS2611', 'TGS2600']
  Shape: (150, 7)

COO1: 70 files
  Columns: ['Time [s]', 'TGS2620', 'TGS2602', 'TGS2603', 'TGS2610', 'TGS2611', 'TGS2600']
  Shape: (150, 7)

COO2: 70 files
  Columns: ['Time [s]', 'TGS2620', 'TGS2602', 'TGS2603', 'TGS2610', 'TGS2611', 'TGS2600']
  Shape: (150, 7)

COO3: 70 files
  Columns: ['Time [s]', 'TGS2620', 'TGS2602', 'TGS2603', 'TGS2610', 'TGS2611', 'TGS2600']
  Shape: (150, 7)

COO4: 70 files
  Columns: ['Time [s]', 'TGS2620', 'TGS2602', 'TGS2603', 'TGS2610', 'TGS2611', 'TGS2600']
  Shape: (150, 7)

COO5: 70 files
  Columns: ['Time [s]', 'TGS2620', 'TGS2602', 'TGS2603', 'TGS2610', 'TGS2611', 'TGS2600']
  Shape: (150, 7)

RO0: 70 files
  Columns: ['Time [s]', 'TGS2620', 'TGS2602', 'TGS2603', 'TGS2610', 'TGS2611', 'TGS2600']
  Shape: (150, 7)

RO1: 70 fil

In [12]:
# Summary of columns consistency
print("\n" + "="*70)
print("SUMMARY OF COLUMN CONSISTENCY ANALYSIS")
print("="*70)

if len(columns_dict) == 1:
    print("\n✓ ALL FILES HAVE CONSISTENT COLUMNS!")
    print(f"\nTotal unique column structure(s): {len(columns_dict)}")
    
    # Display the single column structure
    for columns, oil_info in columns_dict.items():
        print(f"\nColumn names ({len(columns)} columns):")
        for i, col in enumerate(columns, 1):
            print(f"  {i}. {col}")
        
        print(f"\nFiles using this structure ({len(oil_info)} oil types):")
        total_files_with_structure = 0
        for oil_type, first_file, num_files in oil_info:
            print(f"  - {oil_type}: {num_files} files")
            total_files_with_structure += num_files
        print(f"\n  Total files: {total_files_with_structure}")
else:
    print(f"\n✗ FOUND {len(columns_dict)} DIFFERENT COLUMN STRUCTURES!")
    print("\nColumn structures:")
    for i, (columns, oil_info) in enumerate(columns_dict.items(), 1):
        print(f"\n  Structure {i}: {columns}")
        print(f"  Used in: {[info[0] for info in oil_info]}")

if files_with_issues:
    print(f"\n⚠ WARNING: Found {len(files_with_issues)} files with inconsistent columns")
    print("Files with issues:")
    for oil_type, filename in files_with_issues[:10]:  # Show first 10
        print(f"  - {oil_type}/{filename}")
    if len(files_with_issues) > 10:
        print(f"  ... and {len(files_with_issues) - 10} more")
else:
    print("\n✓ All files within each oil type have consistent columns")

print("\n" + "="*70)


SUMMARY OF COLUMN CONSISTENCY ANALYSIS

✓ ALL FILES HAVE CONSISTENT COLUMNS!

Total unique column structure(s): 1

Column names (7 columns):
  1. Time [s]
  2. TGS2620
  3. TGS2602
  4. TGS2603
  5. TGS2610
  6. TGS2611
  7. TGS2600

Files using this structure (19 oil types):
  - CO: 70 files
  - COO: 70 files
  - COO1: 70 files
  - COO2: 70 files
  - COO3: 70 files
  - COO4: 70 files
  - COO5: 70 files
  - RO0: 70 files
  - RO1: 70 files
  - RO2: 70 files
  - RO3: 70 files
  - RO4: 70 files
  - RO5: 70 files
  - SO0: 70 files
  - SO1: 70 files
  - SO2: 70 files
  - SO3: 70 files
  - SO4: 70 files
  - SO5: 70 files

  Total files: 1330

✓ All files within each oil type have consistent columns



In [13]:
# Additional details about data structure
print("\nDATA STRUCTURE DETAILS:")
print("="*70)

# Check data shape consistency
print("\nChecking data shape consistency...")
shape_dict = {}

for oil_folder in oil_folders:
    txt_files = sorted(oil_folder.glob('*.txt'))
    if txt_files:
        df_sample = pd.read_csv(txt_files[0])
        shape = df_sample.shape
        
        if shape not in shape_dict:
            shape_dict[shape] = []
        shape_dict[shape].append(oil_folder.name)

print(f"\nFound {len(shape_dict)} unique data shape(s):")
for shape, oil_types in sorted(shape_dict.items()):
    print(f"  Shape {shape}: {len(oil_types)} oil types")
    if len(oil_types) <= 5:
        print(f"    Oil types: {', '.join(oil_types)}")

# Data type information
print("\nData type information:")
sample_df = pd.read_csv(list(oil_folders[0].glob('*.txt'))[0])
print(sample_df.dtypes)

print("\nSample data (first 5 rows from CO_1.txt):")
print(sample_df.head())


DATA STRUCTURE DETAILS:

Checking data shape consistency...

Found 1 unique data shape(s):
  Shape (150, 7): 19 oil types

Data type information:
Time [s]      int64
TGS2620     float64
TGS2602     float64
TGS2603     float64
TGS2610     float64
TGS2611     float64
TGS2600     float64
dtype: object

Sample data (first 5 rows from CO_1.txt):
   Time [s]   TGS2620   TGS2602   TGS2603   TGS2610   TGS2611   TGS2600
0         1  0.044517  0.030207  0.029896  0.008595  0.018460  0.033513
1         2  0.044510  0.030220  0.029909  0.008594  0.018462  0.033507
2         3  0.044500  0.030234  0.029908  0.008592  0.018458  0.033500
3         4  0.044494  0.030246  0.029910  0.008594  0.018458  0.033490
4         5  0.044485  0.030262  0.029915  0.008593  0.018457  0.033486


## Analysis Results

### Key Findings ✓

1. **Column Standardization: FULLY STANDARDIZED**
   - All 1,330 data files across all 19 oil types use identical columns
   - No inconsistencies found

2. **Column Structure**
   - **7 columns total**: 1 time column + 6 sensor columns
   - Time column: `Time [s]` (integer values, 1-150 seconds)
   - Sensor columns: `TGS2620`, `TGS2602`, `TGS2603`, `TGS2610`, `TGS2611`, `TGS2600` (all float values)

3. **Data Shape Consistency**
   - All files have identical dimensions: **(150 rows, 7 columns)**
   - Each file contains 150 time measurements

4. **Oil Types**
   - 19 different oil types represented
   - Each type has exactly 70 data files
   - Oil types: CO, COO, COO1-5, RO0-5, SO0-5

### Sensor Information
The dataset uses 6 sensors from the TGS series (electronic nose sensors):
- TGS2620: TVOC & smoke detection
- TGS2602: Odor & smoke detection  
- TGS2603: Odor detection
- TGS2610: Alcohol detection
- TGS2611: General air quality
- TGS2600: General air quality

In [14]:
# Statistics about sample counts in each file
print("\n" + "="*70)
print("SAMPLE COUNT STATISTICS")
print("="*70)

file_stats = []

for oil_folder in oil_folders:
    txt_files = sorted(oil_folder.glob('*.txt'))
    
    if txt_files:
        for txt_file in txt_files:
            df = pd.read_csv(txt_file)
            sample_count = len(df)
            file_stats.append({
                'Oil Type': oil_folder.name,
                'File': txt_file.name,
                'Samples': sample_count,
                'Path': str(txt_file)
            })

# Create dataframe for easier analysis
stats_df = pd.DataFrame(file_stats)

# Summary statistics
print("\nOVERALL STATISTICS:")
print(f"  Total files: {len(stats_df)}")
print(f"  Total samples across all files: {stats_df['Samples'].sum():,}")
print(f"  Average samples per file: {stats_df['Samples'].mean():.0f}")
print(f"  Min samples in a file: {stats_df['Samples'].min()}")
print(f"  Max samples in a file: {stats_df['Samples'].max()}")

# By oil type
print("\nSAMPLES BY OIL TYPE:")
oil_type_stats = stats_df.groupby('Oil Type').agg({
    'Samples': ['count', 'sum', 'mean', 'min', 'max']
}).round(0)

for oil_type in sorted(stats_df['Oil Type'].unique()):
    subset = stats_df[stats_df['Oil Type'] == oil_type]
    print(f"\n  {oil_type}:")
    print(f"    Files: {len(subset)}")
    print(f"    Total samples: {subset['Samples'].sum():,}")
    print(f"    Samples per file: {subset['Samples'].iloc[0]} (all files identical)")

# Check if all files have same number of samples
unique_samples = stats_df['Samples'].unique()
if len(unique_samples) == 1:
    print(f"\n✓ ALL FILES HAVE SAME NUMBER OF SAMPLES: {unique_samples[0]} samples per file")
else:
    print(f"\n⚠ Files have DIFFERENT sample counts:")
    for sample_count in sorted(unique_samples):
        count = len(stats_df[stats_df['Samples'] == sample_count])
        print(f"  {sample_count} samples: {count} files")

print("\n" + "="*70)


SAMPLE COUNT STATISTICS

OVERALL STATISTICS:
  Total files: 1330
  Total samples across all files: 199,500
  Average samples per file: 150
  Min samples in a file: 150
  Max samples in a file: 150

SAMPLES BY OIL TYPE:

  CO:
    Files: 70
    Total samples: 10,500
    Samples per file: 150 (all files identical)

  COO:
    Files: 70
    Total samples: 10,500
    Samples per file: 150 (all files identical)

  COO1:
    Files: 70
    Total samples: 10,500
    Samples per file: 150 (all files identical)

  COO2:
    Files: 70
    Total samples: 10,500
    Samples per file: 150 (all files identical)

  COO3:
    Files: 70
    Total samples: 10,500
    Samples per file: 150 (all files identical)

  COO4:
    Files: 70
    Total samples: 10,500
    Samples per file: 150 (all files identical)

  COO5:
    Files: 70
    Total samples: 10,500
    Samples per file: 150 (all files identical)

  RO0:
    Files: 70
    Total samples: 10,500
    Samples per file: 150 (all files identical)

  RO1:


## Summary Statistics

### Sample Count Analysis ✓

**Hoàn toàn đồng nhất về số lượng mẫu!**

- **Tổng số file**: 1,330 files
- **Tổng số mẫu**: 199,500 samples
- **Mẫu/file**: **150 mẫu** (tất cả file đều giống nhau)
- **Mẫu/loại dầu**: 10,500 mẫu (70 files × 150 mẫu)

### Phân bố dữ liệu
- **CO**: 70 files × 150 samples = 10,500 samples
- **COO + COO1-5**: 7 types × 70 files × 150 = 73,500 samples  
- **RO0-5**: 6 types × 70 files × 150 = 63,000 samples
- **SO0-5**: 6 types × 70 files × 150 = 63,000 samples

### Kết luận
✓ Dữ liệu hoàn toàn chuẩn hóa:
- Cột: Giống nhau (7 cột)
- Kích thước: Giống nhau (150 hàng × 7 cột)
- Số file: Đồng đều (70 files/loại dầu)
- Tổng mẫu: 199,500 samples

## Data Cleaning and Train/Validation/Test Split for TimeGAN

Mục tiêu:
- Đọc toàn bộ file `.txt` thành chuỗi thời gian 3D `(n_sequences, seq_len, n_features)`.
- Clean dữ liệu: ép kiểu số, xử lý thiếu dữ liệu, xử lý trùng thời gian, đảm bảo độ dài chuỗi cố định.
- Chia dữ liệu thành 3 tập `train/validation/test` theo tỷ lệ 70/15/15, giữ phân bố theo từng loại dầu.
- Chuẩn hóa MinMax theo **train-only** để tránh data leakage.
- Lưu output ra thư mục `processed_timegan/` để dùng trực tiếp cho TimeGAN.

In [15]:
import numpy as np

# Configuration for TimeGAN prep
RANDOM_SEED = 42
TEST_RATIO = 0.20

FEATURE_COLUMNS = ['TGS2620', 'TGS2602', 'TGS2603', 'TGS2610', 'TGS2611', 'TGS2600']
TIME_COLUMN = 'Time [s]'
EXPECTED_SEQ_LEN = 150

output_dir = dataset_path.parent / 'processed_timegan'
output_dir.mkdir(exist_ok=True)

rng = np.random.default_rng(RANDOM_SEED)

print('Configuration ready.')
print(f'Output directory: {output_dir}')
print(f'Feature columns ({len(FEATURE_COLUMNS)}): {FEATURE_COLUMNS}')
print(f'Test ratio: {TEST_RATIO}')

Configuration ready.
Output directory: /Users/momthu/Documents/duhoc/Hungary/elte/research/oil/processed_timegan
Feature columns (6): ['TGS2620', 'TGS2602', 'TGS2603', 'TGS2610', 'TGS2611', 'TGS2600']
Test ratio: 0.2


In [16]:
# Build metadata of all files
records = []

for oil_folder in sorted([d for d in dataset_path.iterdir() if d.is_dir()]):
    for txt_file in sorted(oil_folder.glob('*.txt')):
        records.append({
            'oil_type': oil_folder.name,
            'file_name': txt_file.name,
            'file_path': str(txt_file)
        })

meta_df = pd.DataFrame(records)

print(f'Total sequences: {len(meta_df)}')
print(f'Oil types: {meta_df["oil_type"].nunique()}')
print('\nSequences per oil type:')
print(meta_df['oil_type'].value_counts().sort_index())

Total sequences: 1330
Oil types: 19

Sequences per oil type:
oil_type
CO      70
COO     70
COO1    70
COO2    70
COO3    70
COO4    70
COO5    70
RO0     70
RO1     70
RO2     70
RO3     70
RO4     70
RO5     70
SO0     70
SO1     70
SO2     70
SO3     70
SO4     70
SO5     70
Name: count, dtype: int64


In [17]:
# Clean one sequence file and return fixed-length feature matrix (seq_len, n_features)
def clean_sequence_file(file_path: str,
                        feature_cols=FEATURE_COLUMNS,
                        time_col=TIME_COLUMN,
                        expected_len=EXPECTED_SEQ_LEN):
    df_raw = pd.read_csv(file_path)

    # Keep only relevant columns and coerce to numeric.
    df = df_raw[[time_col] + feature_cols].copy()
    for col in [time_col] + feature_cols:
        df[col] = pd.to_numeric(df[col], errors='coerce')

    # Sort by time and collapse duplicated timestamps by mean.
    df = df.sort_values(time_col)
    df = df.groupby(time_col, as_index=False)[feature_cols].mean()

    # Reindex to expected timeline and interpolate missing values.
    expected_time = np.arange(1, expected_len + 1)
    df = df.set_index(time_col).reindex(expected_time)
    df[feature_cols] = df[feature_cols].interpolate(method='linear', limit_direction='both')
    df[feature_cols] = df[feature_cols].ffill().bfill()

    # Last fallback for pathological cases.
    if df[feature_cols].isna().any().any():
        df[feature_cols] = df[feature_cols].fillna(0.0)

    return df[feature_cols].to_numpy(dtype=np.float32)


# Build 3D tensor and track cleaning issues
X_list = []
issues = []

for i, row in meta_df.iterrows():
    try:
        x_seq = clean_sequence_file(row['file_path'])
        if x_seq.shape != (EXPECTED_SEQ_LEN, len(FEATURE_COLUMNS)):
            issues.append((row['file_name'], f'unexpected shape {x_seq.shape}'))
        X_list.append(x_seq)
    except Exception as e:
        issues.append((row['file_name'], str(e)))
        X_list.append(np.zeros((EXPECTED_SEQ_LEN, len(FEATURE_COLUMNS)), dtype=np.float32))

X = np.stack(X_list, axis=0)
y = meta_df['oil_type'].to_numpy()

print(f'X shape for TimeGAN: {X.shape} (n_sequences, seq_len, n_features)')
print(f'y shape: {y.shape}')
print(f'Cleaning issues found: {len(issues)}')
if issues:
    print('First 5 issues:')
    for item in issues[:5]:
        print(' -', item)

X shape for TimeGAN: (1330, 150, 6) (n_sequences, seq_len, n_features)
y shape: (1330,)
Cleaning issues found: 0


In [18]:
# Fixed 20% test split and evaluate sequence models with PyTorch
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report
import pandas as pd
import numpy as np

import torch
import torch.nn as nn
from torch.utils.data import Dataset, DataLoader

# Encode labels and keep sequence shape for deep models
le = LabelEncoder()
y_enc = le.fit_transform(y)

# Fixed stratified test split on full sequences
X_train_seq, X_test_seq, y_train, y_test = train_test_split(
    X, y_enc, test_size=TEST_RATIO, stratify=y_enc, random_state=RANDOM_SEED
)

# Split train into train/validation for torch models
X_train_seq, X_val_seq, y_train, y_val = train_test_split(
    X_train_seq,
    y_train,
    test_size=0.15,
    stratify=y_train,
    random_state=RANDOM_SEED,
)

num_classes = len(le.classes_)
seq_len = X.shape[1]
num_features = X.shape[2]

# Dataset helper
class SequenceDataset(Dataset):
    def __init__(self, X_array, y_array):
        self.X = torch.tensor(X_array, dtype=torch.float32)
        self.y = torch.tensor(y_array, dtype=torch.long)

    def __len__(self):
        return len(self.y)

    def __getitem__(self, idx):
        return self.X[idx], self.y[idx]


train_loader = DataLoader(SequenceDataset(X_train_seq, y_train), batch_size=32, shuffle=True)
val_loader = DataLoader(SequenceDataset(X_val_seq, y_val), batch_size=64, shuffle=False)
test_loader = DataLoader(SequenceDataset(X_test_seq, y_test), batch_size=64, shuffle=False)

# Torch device
DEVICE = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
print(f'Using device: {DEVICE}')

# Helper functions

def evaluate_torch_model(model, loader):
    model.eval()
    correct = 0
    total = 0
    all_preds = []
    with torch.no_grad():
        for batch_x, batch_y in loader:
            batch_x = batch_x.to(DEVICE)
            batch_y = batch_y.to(DEVICE)
            logits = model(batch_x)
            preds = logits.argmax(dim=1)
            correct += (preds == batch_y).sum().item()
            total += batch_y.size(0)
            all_preds.extend(preds.cpu().numpy().tolist())
    return correct / total, np.array(all_preds)


def train_torch_model(model, name, train_loader, val_loader, epochs=20, lr=1e-3):
    model.to(DEVICE)
    optimizer = torch.optim.Adam(model.parameters(), lr=lr)
    criterion = nn.CrossEntropyLoss()
    best_state = None
    best_val_acc = -1.0

    for epoch in range(epochs):
        model.train()
        for batch_x, batch_y in train_loader:
            batch_x = batch_x.to(DEVICE)
            batch_y = batch_y.to(DEVICE)

            optimizer.zero_grad()
            logits = model(batch_x)
            loss = criterion(logits, batch_y)
            loss.backward()
            optimizer.step()

        val_acc, _ = evaluate_torch_model(model, val_loader)
        if val_acc > best_val_acc:
            best_val_acc = val_acc
            best_state = {k: v.detach().cpu().clone() for k, v in model.state_dict().items()}

    if best_state is not None:
        model.load_state_dict(best_state)

    test_acc, y_pred = evaluate_torch_model(model, test_loader)
    print(f"{name}: accuracy={test_acc:.4f}")
    return test_acc, y_pred


# PyTorch models
class CNN1DClassifier(nn.Module):
    def __init__(self, num_features, num_classes):
        super().__init__()
        self.net = nn.Sequential(
            nn.Conv1d(num_features, 64, kernel_size=5, padding=2),
            nn.ReLU(),
            nn.BatchNorm1d(64),
            nn.MaxPool1d(2),
            nn.Conv1d(64, 128, kernel_size=3, padding=1),
            nn.ReLU(),
            nn.BatchNorm1d(128),
            nn.AdaptiveAvgPool1d(1),
        )
        self.classifier = nn.Sequential(
            nn.Flatten(),
            nn.Linear(128, 128),
            nn.ReLU(),
            nn.Dropout(0.3),
            nn.Linear(128, num_classes),
        )

    def forward(self, x):
        x = x.transpose(1, 2)
        x = self.net(x)
        return self.classifier(x)


class GRUClassifier(nn.Module):
    def __init__(self, num_features, num_classes):
        super().__init__()
        self.gru = nn.GRU(num_features, 64, batch_first=True)
        self.dropout = nn.Dropout(0.3)
        self.fc = nn.Sequential(
            nn.Linear(64, 64),
            nn.ReLU(),
            nn.Dropout(0.3),
            nn.Linear(64, num_classes),
        )

    def forward(self, x):
        out, _ = self.gru(x)
        out = self.dropout(out[:, -1, :])
        return self.fc(out)


class LSTMClassifier(nn.Module):
    def __init__(self, num_features, num_classes):
        super().__init__()
        self.lstm = nn.LSTM(num_features, 64, batch_first=True)
        self.dropout = nn.Dropout(0.3)
        self.fc = nn.Sequential(
            nn.Linear(64, 64),
            nn.ReLU(),
            nn.Dropout(0.3),
            nn.Linear(64, num_classes),
        )

    def forward(self, x):
        out, _ = self.lstm(x)
        out = self.dropout(out[:, -1, :])
        return self.fc(out)


class CNNGRUClassifier(nn.Module):
    def __init__(self, num_features, num_classes):
        super().__init__()
        self.conv = nn.Sequential(
            nn.Conv1d(num_features, 64, kernel_size=5, padding=2),
            nn.ReLU(),
            nn.MaxPool1d(2),
        )
        self.gru = nn.GRU(64, 64, batch_first=True)
        self.fc = nn.Sequential(
            nn.Linear(64, 64),
            nn.ReLU(),
            nn.Dropout(0.3),
            nn.Linear(64, num_classes),
        )

    def forward(self, x):
        x = x.transpose(1, 2)
        x = self.conv(x)
        x = x.transpose(1, 2)
        out, _ = self.gru(x)
        out = out[:, -1, :]
        return self.fc(out)


# Train/evaluate deep sequence models
results = []
model_specs = [
    ('1D-CNN', CNN1DClassifier(num_features, num_classes)),
    ('GRU', GRUClassifier(num_features, num_classes)),
    ('LSTM', LSTMClassifier(num_features, num_classes)),
    ('CNN-GRU', CNNGRUClassifier(num_features, num_classes)),
]

predictions = {}
for name, model in model_specs:
    acc, y_pred = train_torch_model(model, name, train_loader, val_loader)
    results.append({'model': name, 'accuracy': acc})
    predictions[name] = y_pred

# RandomForest baseline on flattened sequences
X_train_flat = X_train_seq.reshape(X_train_seq.shape[0], -1)
X_test_flat = X_test_seq.reshape(X_test_seq.shape[0], -1)
rf_model = RandomForestClassifier(n_estimators=300, random_state=RANDOM_SEED, n_jobs=-1)
rf_model.fit(X_train_flat, y_train)
y_rf = rf_model.predict(X_test_flat)
rf_acc = accuracy_score(y_test, y_rf)
print(f"RandomForest: accuracy={rf_acc:.4f}")
results.append({'model': 'RandomForest', 'accuracy': rf_acc})

res_df = pd.DataFrame(results).sort_values('accuracy', ascending=False)
print('\nResults:')
print(res_df)

# Show classification report for the best model
best_name = res_df.iloc[0]['model']
if best_name == 'RandomForest':
    y_best = y_rf
else:
    y_best = predictions[best_name]

print('\nClassification report for best model:', best_name)
print(classification_report(y_test, y_best, target_names=le.classes_))

# Save results
res_df.to_csv(output_dir / 'model_results.csv', index=False)
print(f"Saved results to {output_dir / 'model_results.csv'}")

Using device: cpu
1D-CNN: accuracy=0.9211
GRU: accuracy=0.1729
LSTM: accuracy=0.1053
CNN-GRU: accuracy=0.2744
RandomForest: accuracy=1.0000

Results:
          model  accuracy
4  RandomForest  1.000000
0        1D-CNN  0.921053
3       CNN-GRU  0.274436
1           GRU  0.172932
2          LSTM  0.105263

Classification report for best model: RandomForest
              precision    recall  f1-score   support

          CO       1.00      1.00      1.00        14
         COO       1.00      1.00      1.00        14
        COO1       1.00      1.00      1.00        14
        COO2       1.00      1.00      1.00        14
        COO3       1.00      1.00      1.00        14
        COO4       1.00      1.00      1.00        14
        COO5       1.00      1.00      1.00        14
         RO0       1.00      1.00      1.00        14
         RO1       1.00      1.00      1.00        14
         RO2       1.00      1.00      1.00        14
         RO3       1.00      1.00      1.00    